extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/CommonPanacheQueryImpl.java

```diff
@@ -0,0 +1,382 @@
+        page(page.first());
+    }
+
+    // FIXME: what do we do for this one?
+    //    public void lastPage() {
+    //        checkPagination();
+    //        page(page.index(pageCount() - 1));
+    //    }
+
+    public Uni<Boolean> hasNextPage() {
+        checkPagination();
+        return pageCount().map(pageCount -> page.index < (pageCount - 1));
+    }
+
+    public boolean hasPreviousPage() {
+        checkPagination();
+        return page.index > 0;
+    }
+
+    public Uni<Integer> pageCount() {
+        checkPagination();
+        return count().map(count -> {
+            if (count == 0)
+                return 1; // a single page of zero results
+            return (int) Math.ceil((double) count / (double) page.size);
+        });
+    }
+
+    public Page page() {
+        checkPagination();
+        return page;
+    }
+
+    private void checkPagination() {
+        // FIXME: turn into Uni
+        if (page == null) {
+            throw new UnsupportedOperationException("Cannot call a page related method, " +
+                    "call page(Page) or page(int, int) to initiate pagination first");
+        }
+        if (range != null) {
+            throw new UnsupportedOperationException("Cannot call a page related method in a ranged query, " +
+                    "call page(Page) or page(int, int) to initiate pagination first");
+        }
+    }
+
+    public void range(int startIndex, int lastIndex) {
+        this.range = Range.of(startIndex, lastIndex);
+        // reset the page to its default to be able to switch from page to range
+        this.page = null;
+    }
+
+    public void withLock(LockModeType lockModeType) {
+        this.lockModeType = lockModeType;
+    }
+
+    public void withHint(String hintName, Object value) {
+        if (hints == null) {
+            hints = new HashMap<>();
+        }
+        hints.put(hintName, value);
+    }
+
+    // Results
+
+    @SuppressWarnings("unchecked")
+    public Uni<Long> count() {
+        if (AbstractJpaOperations.isNamedQuery(query)) {
+            throw new PanacheQueryException("Unable to perform a count operation on a named query");
+        }
+
+        if (count == null) {
+            count = em.flatMap(em -> {
+                Mutiny.Query<Long> countQuery = em.createQuery(countQuery());
+                if (paramsArrayOrMap instanceof Map)
+                    AbstractJpaOperations.bindParameters(countQuery, (Map<String, Object>) paramsArrayOrMap);
+                else
+                    AbstractJpaOperations.bindParameters(countQuery, (Object[]) paramsArrayOrMap);
+                // FIXME: question about caching the result here
+                return applyFilters(em, () -> countQuery.getSingleResult());
+            });
+        }
+        return count;
+    }
+
+    private String countQuery() {
+        if (countQuery != null) {
+            return countQuery;
+        }
+
+        // try to generate a good count query from the existing query
```

Comment:
All this stuff with string manipulation is extremely dodgy .. :)

Reply:
I suppose you have much of the same in the non-reactive version of Panache? Could we avoid some copy paste for all this code?

Reply:
Yes, good idea.